Fix DOS plot

sofiasanz opened the pull request:
Previously we were expanding the PDOS on a grid as if it was a charge density (using `sisl.DensityMatrix` to expand the PDOS on the realspace grid)  but this was incorrect. With this fix now `sisl.Wavefunction` is used to expand the wavefunction on the grid and then we square the grid to the power of 2 to obtain the PDOS resolved in realspace.

zerothi commented:
Erhm. I think the notation here is somewhat misleading, or perhaps I am just too siesta like.

Generally people talk about PDOS as something projected onto orbitals etc.
However, a wavefunction projection is more like a local DOS, LDOS. So I guess this depends on what you want?

tfrederiksen commented:
Thanks for the fix!

Regarding the wording I agree with @zerothi that when we are on the realspace grid we should talk about LDOS. I am a little confused about `mode="charge"` vs `mode="density"`. Maybe this could be made more concise?

sofiasanz commented:
> Regarding the wording I agree with @zerothi that when we are on the realspace grid we should talk about LDOS.

OK! Thanks for the tip :) I changed PDOS for LDOS. 

> I am a little confused about `mode="charge"` vs `mode="density"`. Maybe this could be made more concise?

Any suggestions? I chose those words to match the charge plots and the density of states, but I'm open to other ideas :)

zerothi commented on hubbard/hamiltonian.py:
```diff
@@ -841,8 +841,8 @@ def DOS(self, egrid, eta=1e-3, spin=[0, 1], dist='Lorentzian', eref=0.):
 
-    def PDOS(self, egrid, eta=1e-3, spin=[0, 1], dist='Lorentzian', eref=0.):
-        """ Obtains the projected density of states (PDOS) of the system with a distribution function
+    def LDOS(self, egrid, eta=1e-3, spin=[0, 1], dist='Lorentzian', eref=0.):
```
I think we mistunderstood each other. What you have here _was_ correct. I.e. that was calculating the `PDOS`. So you can revert this change.

sofiasanz replied:
Oh I see, I will revert this commit. Thanks!

sofiasanz replied:
Done! This last commit was deleted and now HEAD is at  1d62f6519bf0b5f00417ba64d271b33ad2b1efb3

zerothi commented on hubbard/plot/plot.py:
```diff
@@ -249,23 +249,24 @@ def real_space_grid(v, grid_unit, density):
-                D.D[a, a] = v
-                D.density(grid)
-            else:
+            if mode in ['wavefunction', 'density']:
```
this is where it gets blury. I would stick with `mode in ('wavefunction')`. The function `wavefunction` does _not_ give any density, so they shouldn't be confused.

zerothi commented on hubbard/plot/plot.py:
```diff
@@ -249,23 +249,24 @@ def real_space_grid(v, grid_unit, density):
                 else:
                     # In case v is a vector
                     sisl.electron.wavefunction(v, grid, geometry=g)
+            elif mode in ['charge']:
```
I think this is correct, density or charge is good here I think.

zerothi commented on hubbard/plot/plot.py:
```diff
@@ -276,6 +277,9 @@ def real_space_grid(v, grid_unit, density):
         slice_grid = grid.grid[:, :, 0].T.real
 
+        if mode in ['density']:
+            slice_grid = slice_grid**2
```
Hmm, I see what you were trying to do. But I think the arguments are misleading and needs more thinking. When I read `density` I think of charge density of the full system, not a single eigenstate.

sofiasanz replied:
Maybe I can use 'mode=DOS' instead, what do you think?

zerothi replied:
Hmm... No I don't think so. Hmm.. In this case I would probably use `LDOS`.
And `density` and `charge` would be the same. ?

sofiasanz replied:
OK, so use `if mode in ['charge', 'density']` for what we call charge plots (i.e. spinpolarization, charge, chargedif... etc)
and change what we call here mode='density' for mode='LDOS', no? I will do these changes and upload it now.

zerothi replied:
To me that would make more sense, but perhaps @tfrederiksen could chip in...

sofiasanz replied:
Yes, precisely. In the case of `DOS` plots we are squaring a grid that is not computed from the density matrix. The only cases where we use this grid is for charge plots (spin polarization, charge, charge diff, etc.).
For the `DOS` plots we now are using the grid computed from `sisl.wavefunction` and then squaring it to compute the DOS.

tfrederiksen replied:
What about a more generic internal option `square_gridvalues=False/True` to describe explicitly how the passed physical quantity is handled on the real-space grid?

sofiasanz replied:
That is also fine for me, although in this case we would have to use two flags instead of only one (the flag `mode` (or whatever we want to call it) would still be needed to specify whether we need the grid to be computed from the WF or from the DM).

tfrederiksen replied:
Can't the function not simply check internally whether the input is of type WF or DM?

sofiasanz replied:
Using the flag `mode` was my approach to do it :)
The thing is that in many cases the input is simply a vector (it can also be a `sisl.physics.electron.EigenstateElectron` instance e.g. for WF plots) which then is expanded in the real-space grid (where the way in which the real-space grid is computed is determined by the flag `mode` in this case).

zerothi commented on hubbard/plot/spectrum.py:
```diff
@@ -167,18 +167,18 @@ class DOS_distribution(GeometryPlot):
     HH: HubbardHamiltonian
         mean-field Hubbard Hamiltonian
-    PDOS: numpy.ndarray
+    LDOS: numpy.ndarray
```
I think we may have misunderstood each other. This was correct. Your latest commit can be reverted as far as I can tell.

zerothi reviewed (changes requested):
Sorry, I think we misunderstood each other. Your PDOS _was_ correct. It was only the discussion in the MR.

tfrederiksen commented:
We should also revise the class `LDOSmap` which seems to suffer from the same issue.

sofiasanz commented:
> We should also revise the class `LDOSmap` which seems to suffer from the same issue.

The class `LDOSmap` doesn't use the routines from `sisl` to build the real-space grid.

tfrederiksen commented:
> > We should also revise the class `LDOSmap` which seems to suffer from the same issue.
> 
> The class `LDOSmap` doesn't use the routines from `sisl` to build the real-space grid.

OK, you are right! But perhaps the function name is a bit misleading because it does not produce LDOS in the usual sense. Instead it shows the integrated probability density _resolved along a given direction_ in space. Maybe it should be called something else?

sofiasanz commented:
> OK, you are right! But perhaps the function name is a bit misleading because it does not produce LDOS in the usual sense. Instead it shows the integrated probability density _resolved along a given direction_ in space. Maybe it should be called something else?

Agreed. What about `1d2dPDOS` (because the projection can be along the 1D direction or the 2D plane) or `projDOS` (although this may be confused with `PDOS`) or maybe `PlanarAxialDOS`? Or something similar?

tfrederiksen commented:
> Agreed. What about `1d2dPDOS` (because the projection can be along the 1D direction or the 2D plane) or `projDOS` (although this may be confused with `PDOS`) or maybe `PlanarAxialDOS`? Or something similar?

Not sure what would be a good name. We are doing something "unusual" here compared with standard PDOS/LDOS visualizations. I am now also a bit unsure what we had in mind for the `1D` mode. Wouldn't it not be more sensible in that case to compute the actual LDOS on a line (which would be a different calculation)?

More generally, we may want to reconsider the function names (and descriptions) in `spectrum.py`. For instance, we could also change `class Spectrum(Plot)` (not very evident what we have in mind) to something slightly more specific like `Eigenvalue_spectrum(Plot)` to better capture what it does?

sofiasanz commented:
As we discussed @tfrederiksen, I finally changed the approach to having the grid created outside the `plot` class.
Then we just create the grid and pass it to  `plot._realspace_` function to plot it.

I also did some cleaning here and there in the `DOS` plot classes (spectrum.py file) where I added some info and renamed some functions with the aim to clarify a bit what we do.
Also now the function `plot.LDOS` allows to plot the LDOS with a broadening function at a certain desired energy E (committ [ea3fc37](https://github.com/dipc-cc/hubbard/pull/115/commits/ea3fc37978088b486c7a50fabd967fddb296f346)).

tfrederiksen reviewed (changes requested):
Looks good Sofia!

Apart from the comments below, I think there is a problem with the script `test-hubbard-quantitative.py` which calls the old `DOS` function.

tfrederiksen commented on hubbard/plot/spectrum.py:
```diff
@@ -149,36 +144,114 @@ def __init__(self, HubbardHamiltonian, k=[0, 0, 0], spin=0,
+            wf = 0
+            for i in range(WF.shape[1]):
+                v = WF[:,i]
+                wf += v ** 2
```
Not sure it is a good idea to overwrite wf with probability density. It is no longer the same thing.

sofiasanz replied:
It wasn't overwriting it since `wf` is a different variable from `WF`. But I agree that is confusing, I will change the name of `wf` to e.g. `dos`, which is more representative.

sofiasanz commented on hubbard/grid.py:
```diff
@@ -0,0 +1,61 @@
+    # Move geometry within the supercell
+    g = g.move([-xmin, -ymin, -np.amin(g.xyz[:, 2])])
+    # Make z~0 -> z = 0
+    g.xyz[np.where(np.abs(g.xyz[:, 2]) < 1e-3), 2] = 0
```
This is for our own purposes actually, since we typically plot the geometry backbone on top of the grid to have both the LDOS and the geometry in the plot. With this changes we have the same sizes for the boxes of the geometry and the LDOS colormap Of course the user can have their own function to build the grid.

sofiasanz commented on hubbard/plot/spectrum.py:
```diff
@@ -149,36 +144,114 @@ def __init__(self, HubbardHamiltonian, k=[0, 0, 0], spin=0,
 
 
-class DOS_distribution(GeometryPlot):
+class EigenLDOS(GeometryPlot):
```
Yeah it works for me :)

sofiasanz commented on hubbard/plot/spectrum.py:
```diff
@@ -149,36 +144,114 @@ def __init__(self, HubbardHamiltonian, k=[0, 0, 0], spin=0,
+        smearing parameter. Defaults to 1e-3 eV
+    dist: str, optional
+        distribution function. Defaults to Lorentzian distribution
+    eref: float, optional
```
Yes you are right. I will remove this argument. Thanks!

sofiasanz commented on hubbard/grid.py:
```diff
@@ -0,0 +1,61 @@
+        grid = grid.smooth(method='gaussian', r=r_smooth)
+
+    # Slice it to obtain a 2D grid
+    grid = grid.grid[:, :, 0].T.real
```
I think that is better this way for two main reasons: on one hand the object that is being plotted in the imshow plot is not a `sisl.grid` but a 2D `numpy.array`, so I think is clearer in this sense (and actually more versatile). On the other hand, in `LDOS` and `LDOS_from_eigenstates` functions we need the sliced grid to square it to obtain the `LDOS`, and this operation is performed before calling the plotting function, of course.

sofiasanz commented:
@tfrederiksen following your comments and our earlier discussion I made the corresponding changes in commits [7e82723](https://github.com/dipc-cc/hubbard/pull/115/commits/7e827237d006c12688010959e7333de5b130e391), [98d2cad](https://github.com/dipc-cc/hubbard/pull/115/commits/98d2cad56df8cec5d80c5aa5395ea9a9842a306b) and [54cd86d](https://github.com/dipc-cc/hubbard/pull/115/commits/54cd86da805c4dd7ee709de4bb5ec578dad50c58). Let me know what you think!

zerothi commented on hubbard/grid.py:
```diff
@@ -0,0 +1,64 @@
+import sisl
+
+
+def real_space_grid(geometry, vector, shape, xmin, xmax, ymin, ymax, z, mode='wavefunction', **kwargs):
```
The interface here forces it to be 2d with z being the dependent variable. Is that what you want for the future API? In case you want to handle other systems where this is not the natural order, then it might be worthwhile generalizing the interface?

sofiasanz replied:
Hmm maybe I'm mistaken but I think that one can also create a 3D grid with the appropriate `shape`. The `z` variable sets only the origo which is handy if one wants to create a 2D slice of the grid at that particular height.

zerothi replied:
Yes, but my point is that the plot forces a certain direction, not that I have anything against it, but in case you want to generalize this later, it is better to do it now since the API would change.

sofiasanz replied:
OK understood your point, sorry. You are right. I have an idea, will push it in a few minutes :)

zerothi commented on hubbard/plot/spectrum.py:
```diff
@@ -149,36 +144,116 @@ def __init__(self, HubbardHamiltonian, k=[0, 0, 0], spin=0,
+                dos = WF[:,i]
+                grid_i = real_space_grid(self.geometry, dos, kwargs['shape'], xmin, xmax, ymin, ymax, kwargs['z'], mode='wavefunction')
+                # Slice it to obtain a 2D grid
+                slice_grid = grid_i.grid[:, :, 0].T.real
```
I believe this is incorrect?

What you want is to plot the absolute value of the wavefunction, no?
So `grid += np.absolute(grid_i.grid[:, :, 0])`. This is `(real ** 2 - imag ** 2)**0.5`, if you want it squared, then do that.

sofiasanz replied:
Yes thanks!!

zerothi replied:
thanks for correcting my sign mistake! ;)

zerothi commented on hubbard/plot/spectrum.py:
```diff
@@ -196,20 +271,50 @@ def __init__(self, HH, PDOS, sites=[], ext_geom=None, realspace=False, **kwargs)
+                    f = sisl.get_distribution(distribution, smearing=eta, x0=ev[ni])
+                    weight = f(E)
+                    # Slice it to obtain a 2D grid
+                    slice_grid = grid_n.grid[:, :, 0].T.real
```
same here.

zerothi reviewed (commented):
Sorry to jump in here. Just some minor remarks, and a possible bug?

sofiasanz commented:
> Sorry to jump in here. Just some minor remarks, and a possible bug?

Hi Nick! Your comments are always welcome of course! Thank you for the feedback.

sofiasanz commented:
@zerothi  what do you think of the changes in commit [aade059](https://github.com/dipc-cc/hubbard/pull/115/commits/aade059e6836318e5dddcbcea792409473212fed)? Another suitable option is to directly pass the SC to `real_space_grid` instead of its size and origin.

zerothi commented:
> @zerothi what do you think of the changes in commit [aade059](https://github.com/dipc-cc/hubbard/pull/115/commits/aade059e6836318e5dddcbcea792409473212fed)? Another suitable option is to directly pass the SC to `real_space_grid` instead of its size and origin.

Yes, I think this is more future proof. :)

sofiasanz commented:
In commit [42cdcac](https://github.com/dipc-cc/hubbard/pull/115/commits/42cdcac61525b8638cb10879aae67ecc02dbf108) now the SC object is directly passed instead of being created in `real_space_grid`, so now the desired SC (if not the SC defaults to what we had before).
